File: 2/2_cuatrimestre/PDOO/practicas/practica 4/deepSpaceRuby/SpaceCity.rb
require_relative 'SpaceStation'
require_relative 'SpaceCityToUI'

module Deepspace
    class SpaceCity < SpaceStation

        attr_reader :ammoPower, :fuelUnits, :weapons, :hangar, :shieldPower, :name, :nMedals, :pendingDamage, :shieldBoosters

        def initialize(base, rest)
            @collaborators = Array.new(rest)
            @base = base
            newCopy(base)
        end

        def fire
            size = @collaborators.length
            factor = @base.fire

            for i in 0..(size-1)
                factor *= @collaborators[i].fire
            end
            
            return factor
        end

        def protection 
            factor = super

            @collaborators.each do |x|
                factor *= x.protection
            end

            return factor
        end

        def setLoot(loot)
            super(loot)
            return Transformation::NOTRANSFORM
        end

        def getUIversion
            return SpaceCityToUI.new(self)
        end

        def collaborators
           @collaborators 
        end

    end
end

File: 2/2_cuatrimestre/PDOO/practicas/practica 4/deepSpace_4/src/deepspace/SpaceCity.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package deepspace;

import java.util.ArrayList;

/**
 *
 * @author manbolq
 */
public class SpaceCity extends SpaceStation{
    SpaceStation base;
    ArrayList<SpaceStation> collaborators;
    
    
    
    SpaceCity(SpaceStation base, ArrayList<SpaceStation> rest){
        super(base);
        this.base = base;
        this.collaborators = rest;
    }
    
    
    
    
    public ArrayList<SpaceStation> getCollaborators(){
        return collaborators;
    }
    
    
    @Override
    public float fire(){
        float factor = base.fire();
        
        for (SpaceStation station : collaborators){
            factor *= station.fire();
        }
        
        return factor;
    }
    
    
    @Override
    public float protection(){
        float factor = base.protection();
        
        for (SpaceStation station : collaborators){
            factor *= station.protection();
        }
        
        return factor;
    }
    
    
    @Override
    public Transformation setLoot(Loot loot){
        super.setLoot(loot);
        return Transformation.NOTRANSFORM;
    }
    
    @Override
    public SpaceCityToUI getUIversion(){
        return new SpaceCityToUI(this);
    }    
    
}
